### DISCUSSION TITLE
Voyager 1 and 2 mission as one mission dictionary?

### DISCUSSION THREAD

**Discussion by thareUSGS** [pds-data-dictionaries/PDS-Data-Dictionaries.github.io#13] [Q&A]
As discussed [here](https://github.com/pds-data-dictionaries/PDS4-LDD-Issue-Repo/discussions/199), we have a Voyager 1 and 2 migration  PDART coming in. A question was raised about why there are two Voyager 1 and 2 mission dictionaries (namespaces) since they were identical spacecraft.  Thus having a single mission dictionary "Voyager" might make more sense. Essentially both vg1 and vg2 are the same and currently very simple. I assume these have been used by PPI (as created by @jmafi), but I haven't looked very hard.

Should or could we simply have a merged "vgr" mission dictionary?

**Comment by matthewtiscareno** [comment]
My understanding is that the V1 and V2 spacecraft were *not* identical in their particles and fields instruments, which may be why PPI created separate dictionaries.  

Be that as it may, the primary work of migrating Voyager ISS data to PDS4 will be done by RMS, and we have not begun yet.  There is a lot of work that we will do to properly craft the mission dictionary/dictionaries.

**Comment by rgdeen** [comment]
They may not be *identical* but they're close enough that I think separate dictionaries are unwarranted.  It just creates a double-maintenance issue if changes ever have to be made, and would hinder the use of common software to process the bulk of data that IS the same between them.  To the extent there are differences, those parts of the DD can be used with just one spacecraft or the other.

I should note that everything in the DD's currently is in the msn: dictionary and I would strongly suggest using msn: for those items instead of rolling your own.  But there of course will be quite a few VGR-specific items.

**Reply by beatricemueller** [reply]
I am on the team reviewing the msn namespace and this is supposed to be for things common in missions. We question that items relevant only to a specific mission should be in msn as opposed to that specific namespace.

**Comment by rgdeen** [comment]
Exactly @beatricemueller .  The only things *currently* in the VGR dictionaries are mission_phase_name, and spacecraft_clock_start|stop_count.  All of which are already in msn:, and should therefore be used *from* msn.  That's the whole point of msn.  I'm not about to suggest that anything Voyager-specific be added to msn!!

**Comment by beatricemueller** [comment]
@rgdeen Thanks for the clarification! I misunderstood. (Should have looked at the VGR namespaces first).

**Comment by jmafi** [comment]
@rgdeen is correct, all that are in the VG1 and VG2 dictionaries are mission_phase_name and spacecraft_clock_start/stop_count. The creation of these dictionaries predated the existence of the mission commons dictionary (msn). The reason that two were created is that the mission_phase_name permissible values are unique for VG1 and VG2 and so these lists should not be comingled. I'm less certain about the SCLK values. The two spacecraft certain have their own spacecraft clocks, but I don't know whether there is any correlation between their reset an rollover histories. I think that the msn dictionary will be the way to go moving forward, however, we will still need a way of maintaining different mission_phase_name permissible value lists and SCLK format and value specifications for each spacecraft.

**Comment by rgdeen** [comment]
That's the nice thing about msn: ... it doesn't enforce any constraints on mission_phase_name.  I believe there is a way to restrict the values in msn:mission_phase_name for a given mission, but I don't know what that is exactly.

Even so, it doesn't hurt anything to have MPN's that are available to both spacecraft... you simply don't use the ones that don't apply.  The value of having common data elements (thus enabling both cross-mission data mining, and reusable software, not to mention avoiding duplication of the entire DD) far outweighs the marginal advantage of specific validity checking for things like this, IMO.

I should note that in the Mars realm, we take this concept to the extreme.  There actually is NO M2020-specific data dictionary for the bulk of the imagers (a couple do have some very specific things, but most of them do not, including the engineering cameras and Mastcam-Z).  There are NO camera-related mission-specific DD's at all for InSight or MSL (I have a PDART doing MSL data).  The missions are similar enough in terms of the cameras that there is no need for specific DD's for each mission.

There are some differences, but that's okay.  For example, the SCLK epoch for Voyager will be different than that for the Mars missions.  But that's spelled out in mission-specific definition "nuances" for the common keywords.  Even with a different epoch, there are useful multimission properties to SCLK, such as being able to do time sorting, where the variable epoch start doesn't matter.  There's even already a msn:spacecraft_clock_partition to support the rollover on Voyager.  I would support adding something like spacecraft_clock_epoch_time to msn: to make this more explicit.

Anyway, the point is, there is a LOT of mission- or spacecraft- specific variation that can be accommodated by multimission keywords.  I think it makes the metadata far more understandable, mineable, and usable.  And as I said there are ways to impose mission-specific validation constraints that are subsets of the general case... I don't know if they've been used yet, but I know they exist.

I certainly understand msn: not existing yet when these were created.  But I hope we can deprecate these fields, or maybe the entire DDs, and move to a common Voyager DD.  Given that we don't have a mission DD at *all* for the cameras on MSL, InSight, or most of M20, the minor variations between VGR-1 and VGR-2 can be easily accommodated in a single DD and I fervently hope we can do so.  Thanks.

**Reply by jmafi** [reply]
Sorry for the slow reply, I was out of town most of last week and am still catching up.

I like much of what the msn dictionary offers. I do think that there is a lot of value in designing a dictionary which contains parameter that are common to and can be used by multiple missions. However, I'm going to have to strongly disagree with some of what @rgdeen says. The assertion that "it doesn't hurt anything to have MPN's that are available to both spacecraft... you simply don't use the ones that don't apply" is absolutely wrong. MPN's must be restricted to only those which are valid for a given mission, otherwise the attribute becomes meaningless. A Cassini MPN in an InSight data product, for example, is complete nonsense. A VG2 MPN in a VG1 product is equally useless. 

The same problem extends to SCLK values. Older missions like the Voyagers and Galileo do not even have the same format that more recent missions have. Similar to MPNs, listing a SCLK value that has an invalid format or partition number renders the value useless. 

In summary, I like the concept of a common msn dictionary with context-specific (i.e. mission/spacecraft) constraints on multimission attributes which are enforced by the validation tools. The presence of these constraints is absolutely necessary to make the attributes meaningful.

**Comment by rgdeen** [comment]
Okay so you're not seeing the full power of this system @jmafi .

It is possible - I don't know the mechanism, but I know it is possible - to have a common dictionary attribute with, say, 50 valid values and then have a restriction on that attribute _for a given mission only_ to allow only, say, 5 of them.  So the Mission Phase Name at the msn: dictionary level can in fact list every possible MPN for every mission... but then the mission-specific restriction allows you to say only the ones relevant to your mission are valid.  That actually captures the best of both worlds... you can have a common dictionary element while still doing your mission's validation.

The reason I don't know what the mechanism is, is because we haven't bothered to do so in the Mars realm.  But I've talked to Jordan and Myche and others and they all say it is possible.  Personally I think PDS is way over-validated.  Real-world data is messy, and if the spacecraft reports a value that's in principle out of range, we should be able to preserve that value in the labels.  I mean, if we CAN fix it, great, but we can't always do so... and having a value that's out of range could be helpful later, like when someone down the road realizes a bit was stuck on and thus how to recover the actual value, or something like that.  The point is, it is our responsibility to report what the spacecraft told us, unless we have compelling evidence to fix it.  We don't, often.  Especially for older missions - Voyager should really barely have ANY valid values, because the data is so inconsistent.  The data is 40 years old, we have no idea what it "should" be, so we just pass it on in case someone else knows.  And on Mars missions, we're talking multiple terabytes of data a month, and there's no way we can manually clean any of it up.  It would be irresponsible to reject an image (possibly containing THE fossil, you never know) for archive just because a temperature was reported as -300 C.  Believe me, we've been under pressure to do just that when certain label items won't validate because of data range violations.

That said, if you want to strictly validate your data, you're welcome to.  The mechanism allows for it.  I agree that putting in the wrong mission's MPN is meaningless, but that's not going to happen with a competent data provider with software-driven label generation, and so I see little added value in the validator enforcing that.  Bottom line, I almost never see the validator catching real errors - we generally have to make the validation less strict to get data through.  Nevertheless, the mechanism is there, to each his or her own.

In terms of the multimission applicability, it is incredibly useful.  Software doesn't generally need to know what the various MPN's _mean_, it just needs to group/sort them and present the value to the user.  The conceptual meaning of MPN is the same across missions, even if the specific values vary.

Same for SCLK.  A clock is a clock, it ticks at a known rate.  Generic software does not need to know what the SCLK epoch is, or even how fast the clock ticks, in order to sort data in time order by SCLK.  That's the value of using a common keyword, even when the specifics vary.  I have no objection to adding (optional!) fields for what the sclk epoch is, and ticks/second.  That would make it truly multimission, and is probably a good idea.  At the very least, the documentation should say what those are.  But even in the absence of those things, treating all SCLK's the same from the DD standpoint enables multimission software.  We can't keep rewriting every piece of code for every mission.

I do realize that VGR has annoying SCLK partitions due to FDS rollover.  But that's already accounted for in the msn DD.

I hope that helps...

**Comment by jordanpadams** [comment]
FYI. Locking this conversation for the meeting discussion.

**Reply by jordanpadams** [reply]
Thanks for all the input! great start to the in-person discussion
